src/site/asciidoc/manual/json-template-layout.adoc.vm

```diff
@@ -410,6 +410,15 @@ artifact, which contains the following predefined event templates:
   xref:additional-event-template-fields[additional event template fields]
   to avoid `hostName` property lookup at runtime, which incurs an extra cost.)
 
+- https://github.com/apache/logging-log4j2/tree/master/log4j-layout-template-json/src/main/resources/GcpLayout.json[`GcpLayout.json`]
+  described by https://cloud.google.com/logging/docs/structured-logging[Google
+  Cloud Platform structured logging] with additional
+  `_thread`, `_logger` and `_exception` fields. The exception trace, if any,
```

Comment:
@rocketraman, I was exactly working on this PR for the last 1 hour or so. :facepalm: :sweat_smile:

I have a question regarding these extra fields: Why do we have them? Why do we deviate from the standard? Additionally, why do we choose for these fields names?

Reply:
> I have a question regarding these extra fields: Why do we have them?

I thought that having these additional fields would be useful, because one can use Google Cloud Logging to filter on `_thread` and `_logger`, as well as the structured `_exception` field.

> Why do we deviate from the standard?

We're not really -- the Google Cloud logging format specifies certain field names that it explicitly understands, but it also allows for adding additional fields as may be useful.

> Additionally, why do we choose for these fields names?

I prefixed the additional fields with `_` to differentiate them from the "standard" names understood by Google. I actually got this idea from the `GelfLayout.json`. However, I'm also fine with removing the `_`.

Reply:
@rocketraman, would you give me a couple of days to check out how others are incorporating these fields. I don't want to add custom fields only because we just feel like they might be useful. I want to have a little bit more educated guess.

Reply:
Sure take your time to research this. The code for Google's Logback adapter is here:

https://github.com/googleapis/java-logging-logback/blob/master/src/main/java/com/google/cloud/logging/logback/LoggingAppender.java#L296-L327

The use Google's cloud logging API directly ([LogEntry.Builder](https://googleapis.dev/java/google-cloud-logging/latest/com/google/cloud/logging/LogEntry.Builder.html) to create a [LogEntry](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry)), and log:

* timestamp
* message (with stack)
* severity
* labels via the MDC enhancer
* they add the level name, numeric value, and logger name as additional labels
* they have the ability for users to define "enhancers" which can add additional fields

Note that enhancers can write to the payload field, which is arbitrary JSON. When an application writes JSON to standard output as this code would do, the fields that are not standard fields go the payload, along with the standard `message` field.

One approach would be to log just the "standard" fields by default and let users use `EventTemplateAdditionalFields` to add any additional fields they want. However, my personal opinion is that at least the `logger` and `thread` fields are low cost and very useful to include by default.

Reply:
@vy Any updates?

Reply:
@vy To help concretize the end-effect of this template, here are some examples of cloud console log display.

Non-error log: note the `logger` and `thread` in the payload. Personally I like this data in the payload rather than as labels, though one could make an argument for putting logger in `labels` instead.

![image](https://user-images.githubusercontent.com/53049/125824748-2d62aacb-123b-4eee-a491-7c5845c0b36b.png)

And here is an example with an exception -- the destructured exception is in the payload. This is useful because the individual exception fields, especially `exception_class` and `exception_message`, can be filtered on.

![image](https://user-images.githubusercontent.com/53049/125825212-1773ab99-c7e7-48f4-bcc6-c45722a072fc.png)

I thought this would help concretize the end-effect of this template. Looking forward to your reply.

Reply:
Thanks so much for the great work and investigation, I have really appreciated it! I have incorporated your suggestions into #551. Looking forward to your upcoming `log4j-layout-template-json` contributions!